lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -509,6 +509,50 @@ static CallExpr *createContainerKeyedByCall(ASTContext &C, DeclContext *DC,
                                   C.AllocateCopy(argLabels));
 }
 
+/// Looks up the property corresponding to the indicated coding key.
+///
+/// \param conformanceDC The DeclContext we're generating code within.
+/// \param elt The CodingKeys enum case.
+/// \param targetDecl The type to look up properties in.
+///
+/// \return A tuple containing the \c VarDecl for the property, the type that
+/// should be passed when decoding it, and a boolean which is true if
+/// \c encodeIfPresent/\c decodeIfPresent should be used for this property.
+static std::tuple<VarDecl *, Type, bool>
+lookupVarDeclForCodingKeysCase(DeclContext *conformanceDC,
+                               EnumElementDecl *elt,
+                               NominalTypeDecl *targetDecl) {
+  ASTContext &C = elt->getASTContext();
+
+  for (auto decl : targetDecl->lookupDirect(DeclName(elt->getName()))) {
+    if (auto *vd = dyn_cast<VarDecl>(decl)) {
+      if (!vd->isStatic()) {
+        // This is the VarDecl we're looking for.
+
+        auto varType =
+            conformanceDC->mapTypeIntoContext(vd->getValueInterfaceType());
+
+        bool useIfPresentVariant =
+            varType->getAnyNominal() == C.getOptionalDecl();
+
+        if (useIfPresentVariant) {
+          // The type we request out of decodeIfPresent needs to be unwrapped
+          // one level.
+          // e.g. String? => decodeIfPresent(String.self, forKey: ...), not
+          //                 decodeIfPresent(String?.self, forKey: ...)
+          auto boundOptionalType =
+              dyn_cast<BoundGenericType>(varType->getCanonicalType());
+          varType = boundOptionalType->getGenericArgs()[0];
+        }
+
+        return std::make_tuple(vd, varType, useIfPresentVariant);
+      }
+    }
+  }
+
+  llvm_unreachable("Should have found at least 1 var decl");
```

Comment:
I'm assuming this will assert if ever hit?

Reply:
Yes.